### PR TITLE
Improve HostFiltering

### PR DESCRIPTION
**Pull request by benaadams**
Is now included in DefaultBuilder and was showing up in traces

image

TryGetValueFast is addressed separately by https://github.com/aspnet/AspNetCore/pull/9341

Blocked by https://github.com/aspnet/Extensions/pull/1696

### CODE REVIEW

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (benaadams) [inline src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: @Tratcher is there a reason to allow whitespace before and after the host value? Kestrel will trim it and I assume IIS will also?

rfc7230  3.2.4.  Field Parsing says https://tools.ietf.org/html/rfc7230#page-25
```
Optional whitespace occurring before the first non-whitespace octet of the field value or after the last
non-whitespace octet of the field value ought to be excluded by
parsers when extracting the field value from a header field
```
I've changed it to reject rather than trim

Comment (Tratcher) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: It was not making any assumptions about the underlying server. It was valid input so we tried to allow for it. Does this part really show up in your trace?

Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: StringSegment is a 16 byte struct so passed via stack copy
```
Type layout for 'StringSegment'
Size: 16 bytes. Paddings: 0 bytes (%0 of empty space)
|=================================================|
|   0-7: String <Buffer>k__BackingField (8 bytes) |
|-------------------------------------------------|
|  8-11: Int32 <Offset>k__BackingField (4 bytes)  |
|-------------------------------------------------|
| 12-15: Int32 <Length>k__BackingField (4 bytes)  |
|=================================================|
```

`StringSegment.Trim()` => `TrimStart().TrimEnd()`

TrimStart and TrimEnd both fix the string as well as having a loop both of which block inlining; so it gets stack copies for each of them, then stack copied back out for the return; in addition to the fixing cost.

Could change Trim to first check is the start and end are whitespace etc and optimize there; however its either a parser error, or a user error (host set in pipeline with leading/trailing whitespace) and `TryProcessHosts` doesn't trim the hosts in the allow lists so its not a "round-trippable" error as it won't match if used in both places; so seems easiest to reject?

Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: > Does this part really show up in your trace?
```
dotnet new web
dotnet publish -c Release
```
```
 wrk -c 128 -t 4 -d 120 -H "Accept: application/json,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7" -H "Connection: keep-alive" http://127.0.0.1:5000/ -s ./wrk/scripts/pipeline.lua -- 16
```
On sampling it CheckHost+EnsureConfigured shows up as >10% of the rest of the pipeline 

CheckHost+EnsureConfigured: 20,569+1198 = 21,767ms
EndpointRoutingMiddleware onwards: 213,871ms

21,767/213,871 * 100 = 10.2%

image

Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: TrimStart and TrimEnd are players here; though the effect of logging is more significant

image

File: src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs
Comment (Tratcher) [inline src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: Does this maintain/test the distinction between there being no host header and there being an empty host header? The spec makes a distinction.

Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: Hmm... looking into it. Checking what the parser does.

Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: All looks ok (in Kestrel)

I think there is an issue with `Microsoft.AspNetCore.Http.HeaderDictionary`

https://github.com/aspnet/AspNetCore/blob/4649449206edae4992e09419334e736726fd2c8e/src/Http/Http/src/HeaderDictionary.cs#L77-L80

It should be testing an `IsNull` property (which doesn't exist) rather than the `IsNullOrEmpty`?

Comment (Tratcher) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: StringValues doesn't even have an IsNull :-)
The issue you highlight would only be on the setter, it would be fine if the Store already contained the empty value. I think all of the servers implement their own request headers structure, and this isn't an issue for response headers. That just leaves TestServer?

Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: >  That just leaves TestServer?

Yes, but I assume its why the test was inserting `" "` rather than `""` as a value, as empty gets thrown away

Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: Ah, Kestrel does similar thing; but uses `.Count == 0` which I guess would be the equivalent of `.IsNull`

Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: Resolved

**Comment by Tratcher** [comment]
Hmm, that had some upstream side-effects. https://dev.azure.com/dnceng/public/_build/results?buildId=157194&view=ms.vss-test-web.build-test-results-tab

**Comment by anurse** [comment]
Moving to preview 6. Please do not merge until master is open for preview 6 changes. We're locking down preview 5 early due to a tight bit of release scheduling ;)

**Comment by benaadams** [comment]
@Tratcher upstream fixes https://github.com/aspnet/AspNetCore/pull/9359/commits/20dd7e480e94977d3733e744120a90192e33b6cb

**Comment by benaadams** [comment]
Rebased

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (Tratcher) [inline src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: At this point your enforcing a header parsing rule. If there's an issue here it's with the underlying request parsing, not with the client's request. You can remove this whitespace check.

File: src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs
Comment (Tratcher) [inline src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: This is not desirable output. An empty element should be ignored/replaced.

Comment (benaadams) [reply src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: Change `StringValues` to not output empty strings?

Comment (Tratcher) [reply src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: Or Headers.Append to ignore an empty value?

Comment (benaadams) [reply src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: Empty string is different from not present, as you highlighted https://github.com/aspnet/AspNetCore/pull/9359#discussion_r275417254

> Does this maintain/test the distinction between there being no host header and there being an empty host header? The spec makes a distinction.

Comment (Tratcher) [reply src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: A present but empty header has a meaning in the case of Host. Empty values in a header list do not have a meaning and should be suppressed. (e.g. ", , ,,, , , value" is the same as "value"). We do that when writing response headers to the wire, no?

Comment (benaadams) [reply src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: > We do that when writing response headers to the wire, no?

Yes, though empty values can be happily added to the header; which is likely why they are skipped on output.

Will make change to `StringValues` to skip them also for `.ToString()`; if the header has a real value

Comment (benaadams) [reply src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: https://github.com/aspnet/Extensions/pull/1696

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: Dropped

File: src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs
Comment (benaadams) [reply src/Middleware/ResponseCompression/test/BodyWrapperStreamTests.cs]: Changed

**Comment by benaadams** [comment]
Won't pass until aspnet/Extensions#1696 is merged, due to the current behaviour of `StringValues`

**Comment by benaadams** [comment]
Pre
image

Post
image

**Comment by Tratcher** [comment]
Ready to rebase on aspnet/Extensions#1696?

**Comment by benaadams** [comment]
Rebased

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (jkotalik) [inline src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: Why not `0 == (uint)host.Length`?

File: src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs
Comment (jkotalik) [inline src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: @Tratcher I okay if we take this change (to reject whitespace hosts). Thoughts?

**Review by jkotalik — APPROVED** [review]

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: Weird...

Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: Let me recheck here

File: src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs
Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: Dropped this rejection https://github.com/aspnet/AspNetCore/pull/9359#discussion_r282624054 as it should never happen; so should drop the test too

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: Oh, ok. Its because the Jit likes `>=` tests on `uint`s to do bounds check elimination and the next test *was* checking if the first char was space; so it eliminated the first char bounds check.

However, that check is now gone, so can just do a regular zero test.

File: src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs
Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: As the whitespace reject has gone; this test is just checking if whitespace is in the allow list; so it isn't adding anything.

Comment (benaadams) [reply src/Middleware/HostFiltering/test/HostFilteringMiddlewareTests.cs]: Addressed

File: src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs
Comment (benaadams) [reply src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs]: Addressed

**Review by Tratcher — APPROVED** [review]

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
Raised issue for Helix issue https://github.com/aspnet/AspNetCore/issues/11038

**Pull request merged**
